### PR TITLE
Feat/subscribe

### PR DESCRIPTION
**Pull request by rajatsharma**

### CODE REVIEW

File: packages/react-native/android/src/main/java/com/dashx/sdk/DashXData.kt [hunk 1]
```diff
@@ -20,3 +20,15 @@ data class TrackRequest(
 data class TrackResponse(
     val success: Boolean
 )
+
+data class SubscribeRequest(
+    val deviceToken: String,
+    val deviceName: String,
+    val deviceKind: String,
+    val uid: String?,
+    val anonymous_uid: String?
```
Comment (RabidFire) [inline packages/react-native/android/src/main/java/com/dashx/sdk/DashXData.kt:29]: There seems to be some inconsistency with the casing.

File: packages/react-native/index.js [hunk 2]
```diff
@@ -1,6 +1,19 @@
-import { NativeModules } from 'react-native';
+import { NativeEventEmitter, NativeModules } from 'react-native';
 
 const { DashX } = NativeModules;
+const dashXEventEmitter = new NativeEventEmitter(DashX);
+
+const nativeIdentify = DashX.identify;
+
+DashX.identify = (uid, options) => {
```
Comment (RabidFire) [inline packages/react-native/index.js:8]: This function is again redefined at the bottom. Possibly merge conflict issue?

File: packages/react-native/android/src/main/java/com/dashx/sdk/Utils.kt [hunk 3]
```diff
@@ -13,9 +13,9 @@ fun getPrefKey(context: Context) = "$PACKAGE_NAME.$DEFAULT_INSTANCE.$context.pac
 fun getDashXSharedPreferences(context: Context): SharedPreferences = context.getSharedPreferences(getPrefKey(context), Context.MODE_PRIVATE)
 
 @Throws(JSONException::class)
-fun convertMapToJson(readableMap: ReadableMap?): JsonObject? {
-    val iterator = readableMap?.keySetIterator() ?: return null;
+fun convertMapToJson(readableMap: ReadableMap?): JsonObject {
     val jsonObject = JsonObject()
+    val iterator = readableMap!!.keySetIterator()
```
Comment (RabidFire) [inline packages/react-native/android/src/main/java/com/dashx/sdk/Utils.kt:18]: Why did we remove the null check here?

File: packages/react-native/android/src/main/java/com/dashx/sdk/DashXData.kt (hunk 1 above)
Comment (rajatsharma) [reply packages/react-native/android/src/main/java/com/dashx/sdk/DashXData.kt:29]: Renaming

File: packages/react-native/android/src/main/java/com/dashx/sdk/Utils.kt (hunk 3 above)
Comment (rajatsharma) [reply packages/react-native/android/src/main/java/com/dashx/sdk/Utils.kt:18]: Yes, it will be there.

File: packages/react-native/android/src/main/java/com/dashx/sdk/DashXMessagingService.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.dashx.sdk
+
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+
+class DashXMessagingService : FirebaseMessagingService() {
+    private val tag = DashXMessagingService::class.java.simpleName
+    private val dashXClient: DashXClient? = DashXClient.instance
```
Comment (RabidFire) [inline packages/react-native/android/src/main/java/com/dashx/sdk/DashXMessagingService.kt:8]: Remove optional `?`

File: packages/react-native/ios/DashXClient.swift [hunk 5]
```diff
@@ -111,4 +128,24 @@ class DashXClient {
             { error in DashXLog.d(tag: #function, "Encountered an error during track(): \(error)") }
         )
     }
+    
+    func subscribe() {
+        if deviceToken == nil && identityToken == nil {
```
Comment (RabidFire) [inline packages/react-native/ios/DashXClient.swift:133]: In Android, the condition is `||` but here, in iOS, the condition is `&&`. Probably needs to be the former.

File: examples/react-native/index.js [hunk 6]
```diff
@@ -3,19 +3,19 @@
...
 DashX.setLogLevel(0);
 
 DashX.setup({
   publicKey: Config.DASHX_PUBLIC_KEY,
   baseUri: 'http://api.dashx-staging.com/v1',
 });
 
-DashX.identify({ firstName: 'John', lastName: 'Doe' });
-DashX.track('Hello World');
+DashX.setIdentityToken(Config.DASHX_IDENTITY_TOKEN);
```
Comment (RabidFire) [inline examples/react-native/index.js:19]: Please revert this change. We don't want users thinking it's best practice to store an identity token in config.

File: packages/react-native/README.md [hunk 7]
```diff
@@ -18,6 +18,65 @@ $ npm install @dashx/react-native
...
+```gradle
+buildscript {
+  dependencies {
+    // ... other dependencies
+    classpath 'com.google.gms:google-services:4.3.3'
+  }
+}
+```
+
+- And then add this line in your `/android/app/build.gradle`
```
Comment (RabidFire) [inline packages/react-native/README.md:38]: Fix: Add this line in your `/android/app/build.gradle`

File: packages/react-native/README.md [hunk 8]
```diff
@@ -18,6 +18,65 @@ $ npm install @dashx/react-native
...
+
+```gradle
+apply plugin: 'com.google.gms.google-services'
+```
+
+- Add your Android app on Firebase Console.
```
Comment (RabidFire) [inline packages/react-native/README.md:44]: Possible to add a link to Firebase Console?

File: packages/react-native/README.md [hunk 9]
```diff
@@ -18,6 +18,65 @@ $ npm install @dashx/react-native
...
+
+- Add these lines in your `/ios/{projectName}/AppDelegate.m`
+
+  - At the top of the file import Firebase:
+
+  ```objective-c
+  #import <Firebase.h>
+  ```
+
+  - And inside your `didFinishLaunchingWithOptions` add this:
```
Comment (RabidFire) [inline packages/react-native/README.md]: Fix: Inside your `didFinishLaunchingWithOptions`, add this:

File: packages/react-native/README.md [hunk 10]
```diff
@@ -18,6 +18,65 @@ $ npm install @dashx/react-native
...
+
+  ```objective-c
+  if ([FIRApp defaultApp] == nil) {
+    [FIRApp configure];
+  }
+  ```
+
+- In your pod file add this:
```
Comment (RabidFire) [inline packages/react-native/README.md:66]: Fix: In your `Podfile`, add this:

File: packages/react-native/README.md [hunk 11]
```diff
@@ -78,3 +137,20 @@ DashX.track('event_name', { hello: 'world' } /* Event data */);
 ```javascript
 DashX.reset();
 ```
+
+### Set Identity Token
+
+In order to subcribe to push notifications you need to set identity token like so
+
+```javascript
+DashX.setIdentityToken('your_public_key');
```
Comment (RabidFire) [inline packages/react-native/README.md:144]: The placeholder text ('your_public_key') is misleading. This section needs more explanation.

File: packages/react-native/README.md [hunk 12]
```diff
@@ -18,6 +18,65 @@ $ npm install @dashx/react-native
...
+
+```gradle
+apply plugin: 'com.google.gms.google-services'
+```
+
+- Add your Android app on Firebase Console.
+
+- Download `google-services.json` from there.
+
+- Add `google-services.json` at the following location `/android/app/google-services.json`
```
Comment (RabidFire) [inline packages/react-native/README.md:48]: Fix: Add `google-services.json` at the following location: `/android/app/google-services.json`

File: packages/react-native/README.md [hunk 13]
```diff
@@ -18,6 +18,65 @@ $ npm install @dashx/react-native
...
+
+### Setup for ios
```
Comment (RabidFire) [inline packages/react-native/README.md:50]: Fix: Setup for iOS

File: packages/react-native/README.md [hunk 14]
```diff
@@ -18,6 +18,63 @@ $ npm install @dashx/react-native
...
+}
+```
+
+- In your pod file add this:
+
+```ruby
+pod 'FirebaseInstanceID', :modular_headers => true
+```
+
+- Add your ios app on Firebase Console.
```
Comment (RabidFire) [inline packages/react-native/README.md:72]: Fix `iOS` capitalization here as well. Add link to Firebase Console if possible.

File: packages/react-native/README.md (hunk 11 above)
Comment (rajatsharma) [reply packages/react-native/README.md:144]: Ah, this should be `user's_identity_token`
